compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -750,8 +750,16 @@ private boolean isAssignableMapType(BMapType sourceMapType, BRecordType targetRe
         }
 
         for (BField field : targetRecType.fields.values()) {
-            if (!(Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL) &&
-                    isAssignable(sourceMapType.constraint, field.type))) {
+            if (!Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL)) {
+                return false;
+            }
+
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.READONLY) &&
+                    !Symbols.isFlagOn(sourceMapType.flags, Flags.READONLY)) {
```

Comment:
Should we consider extracting this out as a method? Seeing this repeated a couple of times + would probably be easier to read when used in more complex checks